<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Cryptocurrency Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/BNBBUSD_trades_1m.csv" target="_blank">https://1824251045.rsc.cdn77.org/web/algohouse/data/BNBBUSD_trades_1m.csv</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


you are running the lab in your  browser, so we will install the libraries using `piplite`


In [1]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [2]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [3]:
# import pandas library
import pandas as pd
import numpy as np
pd.set_option("display.precision", 2)
pd.options.display.float_format = '{:.2f}'.format



This function will download the dataset into your browser


In [4]:
#This function will download the dataset into your browser 

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

You can also assign the dataset to any variable you create.

</p>


In [5]:
path = "https://1824251045.rsc.cdn77.org/web/algohouse/data/BNBBUSD_trades_1m.csv"

you will need to download the dataset; if you are running locally, please comment out the following


In [6]:
#you will need to download the dataset; if you are running locally, please comment out the following 
await download(path, "BNBBUSD_trades_1m.csv")
path="BNBBUSD_trades_1m.csv"

This dataset was hosted on IBM Cloud object. Click <a href="https://cocl.us/DA101EN_object_storage?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01">HERE</a> for free storage.


In [7]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"

df = pd.read_csv(path,low_memory=False, index_col=0)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [8]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


ts   open   high    low  close  volume  rec_count  \
0  2022-11-11 14:38:00 284.08 284.08 282.31 282.31 2047.23         62   
1  2022-11-11 14:39:00 282.22 282.42 281.61 281.73  791.29         67   
2  2022-11-11 14:40:00 281.73 282.33 281.23 281.71  734.08         60   
3  2022-11-11 14:41:00 281.45 281.61 280.73 281.59 1855.34         66   
4  2022-11-11 14:42:00 281.72 282.25 281.38 281.38 2878.22         61   

   avg_price  
0     283.24  
1     282.08  
2     281.68  
3     281.22  
4     281.91

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [9]:
# Write your code below and press Shift+Enter to execute 

<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Add Headers</h3>
<p>
Take a look at our dataset. Pandas automatically set the header with an integer starting from 0.
</p>

<p>
Thus, we have to add headers manually.
</p>
<p>
First, we create a list "headers" that include all column names in order.
Then, we use <code>dataframe.columns = headers</code> to replace the headers with the list we created.
</p>


In [10]:
# create headers list
headers = ["Ts","Open","High","Low","Close","Volume","Rec_count","Avg_price"]
print("headers\n", headers)

headers
 ['Ts', 'Open', 'High', 'Low', 'Close', 'Volume', 'Rec_count', 'Avg_price']


We replace headers and recheck our dataframe:


In [11]:
df.columns = headers
df.head(10)

Ts   Open   High    Low  Close  Volume  Rec_count  \
0  2022-11-11 14:38:00 284.08 284.08 282.31 282.31 2047.23         62   
1  2022-11-11 14:39:00 282.22 282.42 281.61 281.73  791.29         67   
2  2022-11-11 14:40:00 281.73 282.33 281.23 281.71  734.08         60   
3  2022-11-11 14:41:00 281.45 281.61 280.73 281.59 1855.34         66   
4  2022-11-11 14:42:00 281.72 282.25 281.38 281.38 2878.22         61   
5  2022-11-11 14:43:00 281.34 282.17 281.27 281.71  232.52         52   
6  2022-11-11 14:44:00 281.74 282.41 281.74 282.41  225.26         44   
7  2022-11-11 14:45:00 282.54 285.49 282.54 285.48 4273.81         87   
8  2022-11-11 14:46:00 285.98 285.98 284.34 285.94 2158.12         83   
9  2022-11-11 14:47:00 285.97 293.00 285.96 290.49 6091.80        101   

   Avg_price  
0     283.24  
1     282.08  
2     281.68  
3     281.22  
4     281.91  
5     281.84  
6     282.23  
7     284.17  
8     285.06  
9     289.20

We can remove the missing values using dropna() 


We can drop missing values along the column "open" as follows:


In [20]:
df=df.dropna(subset=["Open"], axis=0)
df.head(5)

Ts   Open   High    Low  Close  Volume  Rec_count  \
0  2022-11-11 14:38:00 284.08 284.08 282.31 282.31 2047.23         62   
1  2022-11-11 14:39:00 282.22 282.42 281.61 281.73  791.29         67   
2  2022-11-11 14:40:00 281.73 282.33 281.23 281.71  734.08         60   
3  2022-11-11 14:41:00 281.45 281.61 280.73 281.59 1855.34         66   
4  2022-11-11 14:42:00 281.72 282.25 281.38 281.38 2878.22         61   

   Avg_price  
0     283.24  
1     282.08  
2     281.68  
3     281.22  
4     281.91

Now, we have successfully read the raw dataset and added the correct headers into the dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [13]:
# Write your code below and press Shift+Enter to execute 

<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>bnb.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [14]:
# check the data type of data frame "df" by .dtypes
df.dtypes
#If it doesnt work use, you can use  "print()" function


Ts            object
Open         float64
High         float64
Low          float64
Close        float64
Volume       float64
Rec_count      int64
Avg_price    float64
dtype: object

A series with the data type of each column is returned.


<p>
As shown above, it is clear to see that the data type of "open" and "high" is <code>float64</code>, "ts" is object <code>object</code>, etc.
</p>
<p>
These data types can be changed; we will learn how to accomplish this in a later module.
</p>


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [15]:
df.describe()

Open     High      Low    Close   Volume  Rec_count  Avg_price
count 67212.00 67212.00 67212.00 67212.00 67212.00   67212.00   67212.00
mean    273.37   273.48   273.26   273.37   342.64      20.02     273.37
std      21.28    21.28    21.27    21.28   779.06      14.93      21.28
min     219.66   220.37   219.13   219.70     0.02       1.00     219.63
25%     249.15   249.27   249.05   249.16    56.92      10.00     249.15
50%     275.84   276.00   275.72   275.84   143.88      16.00     275.85
75%     290.37   290.46   290.30   290.37   345.47      26.00     290.37
max     317.21   317.66   316.65   317.20 36434.32     114.00     317.32

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "id" has 67212 counts, the mean value of this column is 33605.5, the standard deviation is 19402.6, the minimum value is 0, 25th percentile is 16802, 50th percentile is 33605, 75th percentile is 50408, and the maximum value is 67211. <br>

However, what if we would also like to check all the columns including those that are of type object? <br><br>

You can add an argument <code>include = "all"</code> inside the bracket. Let's try it again.

</p>


In [16]:
# describe all the columns in "df" 
df.describe(include = "all")

Ts     Open     High      Low    Close   Volume  \
count                 67212 67212.00 67212.00 67212.00 67212.00 67212.00   
unique                67212      NaN      NaN      NaN      NaN      NaN   
top     2022-11-11 14:38:00      NaN      NaN      NaN      NaN      NaN   
freq                      1      NaN      NaN      NaN      NaN      NaN   
mean                    NaN   273.37   273.48   273.26   273.37   342.64   
std                     NaN    21.28    21.28    21.27    21.28   779.06   
min                     NaN   219.66   220.37   219.13   219.70     0.02   
25%                     NaN   249.15   249.27   249.05   249.16    56.92   
50%                     NaN   275.84   276.00   275.72   275.84   143.88   
75%                     NaN   290.37   290.46   290.30   290.37   345.47   
max                     NaN   317.21   317.66   316.65   317.20 36434.32   

        Rec_count  Avg_price  
count    67212.00   67212.00  
unique        NaN        NaN  
top           NaN        NaN  
freq          NaN        NaN  
mean        20.02     273.37  
std         14.93      21.28  
min          1.00     219.63  
25%         10.00     249.15  
50%         16.00     275.85  
75%         26.00     290.37  
max        114.00     317.32

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'length' and 'compression-ratio'.

</div>


In [17]:
# Write your code below and press Shift+Enter to execute 

<details><summary>Click here for the solution</summary>

```python
df[['ts', 'volume']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [18]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 67212 entries, 0 to 67211
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Ts         67212 non-null  object 
 1   Open       67212 non-null  float64
 2   High       67212 non-null  float64
 3   Low        67212 non-null  float64
 4   Close      67212 non-null  float64
 5   Volume     67212 non-null  float64
 6   Rec_count  67212 non-null  int64  
 7   Avg_price  67212 non-null  float64
dtypes: float64(6), int64(1), object(1)
memory usage: 4.4+ MB


<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                                         |
| ----------------- | ------- | ---------- | ---------------------------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                                     |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'                       |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe                   |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab                         |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
